examples/shouldi/shouldi/cargo_audit.py

```diff
@@ -0,0 +1,56 @@
+import json
+import asyncio
+from typing import Dict, Any
+
+from dffml.df.base import op
+from dffml.df.types import Definition
+
+package_src_dir = Definition(name="package_src_dir", primitive="str")
+cargo_audit_output = Definition(
+    name="golangci_lint_output", primitive="Dict[str, Any]"
+)
+
+
+class CargoAuditError(Exception):
+    """
+    Raised when cargo-audit fails
+    """
+
+
+async def run_cargo_build(pkg_input: str):
+
+    new_proc = await asyncio.create_subprocess_exec(
+        "cargo",
+        "build",
+        "--release",
+        cwd=pkg_input,
+        stdout=asyncio.subprocess.PIPE,
+        stderr=asyncio.subprocess.PIPE,
+    )
+    stdout, stderr = await new_proc.communicate()
+    if stderr:
```

Comment:
```suggestion
    if new_proc.returncode != 0:
```

```
2020-04-03T15:58:34.8226674Z ======================================================================
2020-04-03T15:58:34.8226851Z ERROR: test_run (tests.test_cargo_audit.TestRunCargoAuditOp)
2020-04-03T15:58:34.8227630Z ----------------------------------------------------------------------
2020-04-03T15:58:34.8227750Z Traceback (most recent call last):
2020-04-03T15:58:34.8229432Z   File "/home/runner/work/dffml/dffml/dffml/util/asynctestcase.py", line 66, in run_it
2020-04-03T15:58:34.8229875Z     result = self.loop.run_until_complete(coro(*args, **kwargs))
2020-04-03T15:58:34.8230146Z   File "/opt/hostedtoolcache/Python/3.7.6/x64/lib/python3.7/asyncio/base_events.py", line 583, in run_until_complete
2020-04-03T15:58:34.8230305Z     return future.result()
2020-04-03T15:58:34.8230598Z   File "/home/runner/work/dffml/dffml/dffml/util/net.py", line 209, in wrapper
2020-04-03T15:58:34.8230792Z     return await func(*(list(args) + [directory_path]), **kwds)
2020-04-03T15:58:34.8230981Z   File "/home/runner/work/dffml/dffml/dffml/util/net.py", line 209, in wrapper
2020-04-03T15:58:34.8231258Z     return await func(*(list(args) + [directory_path]), **kwds)
2020-04-03T15:58:34.8231465Z   File "/home/runner/work/dffml/dffml/dffml/util/net.py", line 209, in wrapper
2020-04-03T15:58:34.8231696Z     return await func(*(list(args) + [directory_path]), **kwds)
2020-04-03T15:58:34.8231877Z   File "/home/runner/work/dffml/dffml/examples/shouldi/tests/test_cargo_audit.py", line 37, in test_run
2020-04-03T15:58:34.8232471Z     await run_cargo_build(cargo_audit / "cargo-audit-0.11.2")
2020-04-03T15:58:34.8232825Z   File "/home/runner/work/dffml/dffml/examples/shouldi/shouldi/cargo_audit.py", line 32, in run_cargo_build
2020-04-03T15:58:34.8233031Z     raise Exception(stderr.decode())
2020-04-03T15:58:34.8233262Z Exception:  Downloading crates ...
2020-04-03T15:58:34.8233385Z   Downloaded serde v1.0.104
2020-04-03T15:58:34.8233538Z   Downloaded serde_json v1.0.46
2020-04-03T15:58:34.8233658Z   Downloaded thiserror v1.0.10
2020-04-03T15:58:34.8233881Z   Downloaded abscissa_core v0.5.2
2020-04-03T15:58:34.8234002Z   Downloaded rustsec v0.18.0
2020-04-03T15:58:34.8234221Z   Downloaded home v0.5.3
2020-04-03T15:58:34.8234349Z   Downloaded gumdrop v0.7.0
2020-04-03T15:58:34.8234486Z   Downloaded serde_derive v1.0.104
2020-04-03T15:58:34.8234627Z   Downloaded itoa v0.4.4
2020-04-03T15:58:34.8234977Z   Downloaded thiserror-impl v1.0.10
2020-04-03T15:58:34.8235233Z   Downloaded ryu v1.0.2
2020-04-03T15:58:34.8235586Z   Downloaded wait-timeout v0.2.0
2020-04-03T15:58:34.8235711Z   Downloaded once_cell v1.3.1
2020-04-03T15:58:34.8235979Z   Downloaded signal-hook v0.1.12
2020-04-03T15:58:34.8236086Z   Downloaded semver v0.9.0
2020-04-03T15:58:34.8236545Z   Downloaded semver-parser v0.9.0
2020-04-03T15:58:34.8236805Z   Downloaded secrecy v0.6.0
2020-04-03T15:58:34.8236992Z   Downloaded libc v0.2.66
2020-04-03T15:58:34.8237448Z   Downloaded backtrace v0.3.42
2020-04-03T15:58:34.8237962Z   Downloaded crates-index v0.13.4
2020-04-03T15:58:34.8238411Z   Downloaded cargo-lock v4.0.1
2020-04-03T15:58:34.8238671Z   Downloaded git2 v0.11.0
2020-04-03T15:58:34.8239396Z   Downloaded tracing-subscriber v0.1.6
2020-04-03T15:58:34.8239662Z   Downloaded platforms v0.2.1
2020-04-03T15:58:34.8239989Z   Downloaded canonical-path v2.0.2
2020-04-03T15:58:34.8240240Z   Downloaded chrono v0.4.10
2020-04-03T15:58:34.8240383Z   Downloaded cvss v1.0.0
2020-04-03T15:58:34.8240526Z   Downloaded tracing v0.1.12
2020-04-03T15:58:34.8240644Z   Downloaded abscissa_derive v0.5.0
2020-04-03T15:58:34.8241061Z   Downloaded generational-arena v0.2.7
2020-04-03T15:58:34.8241501Z   Downloaded tracing-log v0.1.1
2020-04-03T15:58:34.8241945Z   Downloaded color-backtrace v0.3.0
2020-04-03T15:58:34.8242204Z   Downloaded regex v1.3.3
2020-04-03T15:58:34.8242348Z   Downloaded gumdrop_derive v0.7.0
2020-04-03T15:58:34.8242662Z   Downloaded proc-macro2 v1.0.8
2020-04-03T15:58:34.8243248Z   Downloaded semver-parser v0.7.0
2020-04-03T15:58:34.8243438Z   Downloaded darling v0.10.2
2020-04-03T15:58:34.8243849Z   Downloaded backtrace-sys v0.1.32
2020-04-03T15:58:34.8244097Z   Downloaded url v2.1.1
2020-04-03T15:58:34.8244669Z   Downloaded openssl-probe v0.1.2
2020-04-03T15:58:34.8245156Z   Downloaded error-chain v0.12.1
2020-04-03T15:58:34.8245444Z   Downloaded smallvec v0.6.13
2020-04-03T15:58:34.8245682Z   Downloaded petgraph v0.5.0
2020-04-03T15:58:34.8245906Z   Downloaded quote v1.0.2
2020-04-03T15:58:34.8246333Z   Downloaded openssl-sys v0.9.53
2020-04-03T15:58:34.8246803Z   Downloaded num-integer v0.1.42
2020-04-03T15:58:34.8247268Z   Downloaded libgit2-sys v0.10.0
2020-04-03T15:58:34.8247735Z   Downloaded signal-hook-registry v1.2.0
2020-04-03T15:58:34.8248191Z   Downloaded tracing-core v0.1.9
2020-04-03T15:58:34.8248474Z   Downloaded ident_case v1.0.1
2020-04-03T15:58:34.8248714Z   Downloaded owning_ref v0.4.0
2020-04-03T15:58:34.8249190Z   Downloaded tracing-attributes v0.1.6
2020-04-03T15:58:34.8249699Z   Downloaded rustc-demangle v0.1.16
2020-04-03T15:58:34.8249976Z   Downloaded synstructure v0.12.3
2020-04-03T15:58:34.8250230Z   Downloaded matchers v0.0.1
2020-04-03T15:58:34.8250478Z   Downloaded time v0.1.42
2020-04-03T15:58:34.8250939Z   Downloaded num-traits v0.2.11
2020-04-03T15:58:34.8251226Z   Downloaded zeroize v1.1.0
2020-04-03T15:58:34.8251477Z   Downloaded syn v1.0.14
2020-04-03T15:58:34.8251725Z   Downloaded autocfg v1.0.0
2020-04-03T15:58:34.8251965Z   Downloaded matches v0.1.8
2020-04-03T15:58:34.8252420Z   Downloaded percent-encoding v2.1.0
2020-04-03T15:58:34.8252706Z   Downloaded stable_deref_trait v1.1.1
2020-04-03T15:58:34.8252939Z   Downloaded fixedbitset v0.2.0
2020-04-03T15:58:34.8253185Z   Downloaded memchr v2.3.0
2020-04-03T15:58:34.8253428Z   Downloaded darling_core v0.10.2
2020-04-03T15:58:34.8253877Z   Downloaded libssh2-sys v0.2.14
2020-04-03T15:58:34.8254361Z   Downloaded libz-sys v1.0.25
2020-04-03T15:58:34.8254654Z   Downloaded autocfg v0.1.7
2020-04-03T15:58:34.8255109Z   Downloaded pkg-config v0.3.17
2020-04-03T15:58:34.8255403Z   Downloaded idna v0.2.0
2020-04-03T15:58:34.8257207Z   Downloaded indexmap v1.3.1
2020-04-03T15:58:34.8257765Z   Downloaded arc-swap v0.4.4
2020-04-03T15:58:34.8258084Z   Downloaded version_check v0.1.5
2020-04-03T15:58:34.8258523Z   Downloaded maybe-uninit v2.0.0
2020-04-03T15:58:34.8258989Z   Downloaded regex-syntax v0.6.13
2020-04-03T15:58:34.8259459Z   Downloaded regex-automata v0.1.8
2020-04-03T15:58:34.8259742Z   Downloaded darling_macro v0.10.2
2020-04-03T15:58:34.8259980Z   Downloaded fnv v1.0.6
2020-04-03T15:58:34.8260403Z   Downloaded unicode-bidi v0.3.4
2020-04-03T15:58:34.8260682Z   Downloaded strsim v0.9.3
2020-04-03T15:58:34.8261135Z   Downloaded unicode-normalization v0.1.12
2020-04-03T15:58:34.8261402Z   Downloaded byteorder v1.3.2
2020-04-03T15:58:34.8261822Z   Downloaded utf8-ranges v1.0.4
2020-04-03T15:58:34.8262103Z   Downloaded smallvec v1.1.0
2020-04-03T15:58:34.8262502Z   Downloaded jobserver v0.1.19
2020-04-03T15:58:34.8263011Z   Downloaded aho-corasick v0.7.6
2020-04-03T15:58:34.8263298Z    Compiling libc v0.2.66
2020-04-03T15:58:34.8263728Z    Compiling proc-macro2 v1.0.8
2020-04-03T15:58:34.8264331Z    Compiling unicode-xid v0.2.0
2020-04-03T15:58:34.8264769Z    Compiling syn v1.0.14
2020-04-03T15:58:34.8264999Z    Compiling autocfg v1.0.0
2020-04-03T15:58:34.8265454Z    Compiling pkg-config v0.3.17
2020-04-03T15:58:34.8265744Z    Compiling serde v1.0.104
2020-04-03T15:58:34.8266169Z    Compiling cfg-if v0.1.10
2020-04-03T15:58:34.8266451Z    Compiling autocfg v0.1.7
2020-04-03T15:58:34.8266691Z    Compiling log v0.4.8
2020-04-03T15:58:34.8266930Z    Compiling lazy_static v1.4.0
2020-04-03T15:58:34.8267273Z    Compiling matches v0.1.8
2020-04-03T15:58:34.8267867Z    Compiling smallvec v1.1.0
2020-04-03T15:58:34.8269085Z    Compiling byteorder v1.3.2
2020-04-03T15:58:34.8269448Z    Compiling rustc-demangle v0.1.16
2020-04-03T15:58:34.8269553Z    Compiling memchr v2.3.0
2020-04-03T15:58:34.8269676Z    Compiling bitflags v1.2.1
2020-04-03T15:58:34.8269945Z    Compiling maybe-uninit v2.0.0
2020-04-03T15:58:34.8270208Z    Compiling percent-encoding v2.1.0
2020-04-03T15:58:34.8270465Z    Compiling regex-syntax v0.6.13
2020-04-03T15:58:34.8270592Z    Compiling version_check v0.1.5
2020-04-03T15:58:34.8270705Z    Compiling strsim v0.9.3
2020-04-03T15:58:34.8270805Z    Compiling ryu v1.0.2
2020-04-03T15:58:34.8270921Z    Compiling ident_case v1.0.1
2020-04-03T15:58:34.8271035Z    Compiling fnv v1.0.6
2020-04-03T15:58:34.8271295Z    Compiling utf8-ranges v1.0.4
2020-04-03T15:58:34.8271545Z    Compiling semver-parser v0.7.0
2020-04-03T15:58:34.8271802Z    Compiling arc-swap v0.4.4
2020-04-03T15:58:34.8271918Z    Compiling fixedbitset v0.2.0
2020-04-03T15:58:34.8272019Z    Compiling itoa v0.4.4
2020-04-03T15:58:34.8272133Z    Compiling stable_deref_trait v1.1.1
2020-04-03T15:58:34.8272394Z    Compiling openssl-probe v0.1.2
2020-04-03T15:58:34.8272515Z    Compiling termcolor v1.1.0
2020-04-03T15:58:34.8272635Z    Compiling glob v0.3.0
2020-04-03T15:58:34.8272750Z    Compiling zeroize v1.1.0
2020-04-03T15:58:34.8272849Z    Compiling ansi_term v0.11.0
2020-04-03T15:58:34.8272962Z    Compiling home v0.5.3
2020-04-03T15:58:34.8273079Z    Compiling once_cell v1.3.1
2020-04-03T15:58:34.8273349Z    Compiling semver-parser v0.9.0
2020-04-03T15:58:34.8273604Z    Compiling canonical-path v2.0.2
2020-04-03T15:58:34.8273861Z    Compiling num-traits v0.2.11
2020-04-03T15:58:34.8274121Z    Compiling num-integer v0.1.42
2020-04-03T15:58:34.8274225Z    Compiling indexmap v1.3.1
2020-04-03T15:58:34.8274488Z    Compiling generational-arena v0.2.7
2020-04-03T15:58:34.8274745Z    Compiling tracing-core v0.1.9
2020-04-03T15:58:34.8274861Z    Compiling thread_local v1.0.1
2020-04-03T15:58:34.8275115Z    Compiling unicode-bidi v0.3.4
2020-04-03T15:58:34.8275383Z    Compiling unicode-normalization v0.1.12
2020-04-03T15:58:34.8275628Z    Compiling error-chain v0.12.1
2020-04-03T15:58:34.8275747Z    Compiling owning_ref v0.4.0
2020-04-03T15:58:34.8275866Z    Compiling idna v0.2.0
2020-04-03T15:58:34.8275984Z    Compiling quote v1.0.2
2020-04-03T15:58:34.8276098Z    Compiling jobserver v0.1.19
2020-04-03T15:58:34.8276209Z    Compiling time v0.1.42
2020-04-03T15:58:34.8276484Z    Compiling signal-hook-registry v1.2.0
2020-04-03T15:58:34.8276589Z    Compiling atty v0.2.14
2020-04-03T15:58:34.8276846Z    Compiling wait-timeout v0.2.0
2020-04-03T15:58:34.8277101Z    Compiling tracing-log v0.1.1
2020-04-03T15:58:34.8277216Z    Compiling url v2.1.1
2020-04-03T15:58:34.8277473Z    Compiling regex-automata v0.1.8
2020-04-03T15:58:34.8277593Z    Compiling smallvec v0.6.13
2020-04-03T15:58:34.8277847Z    Compiling aho-corasick v0.7.6
2020-04-03T15:58:34.8277949Z    Compiling cc v1.0.50
2020-04-03T15:58:34.8278206Z    Compiling signal-hook v0.1.12
2020-04-03T15:58:34.8278322Z    Compiling petgraph v0.5.0
2020-04-03T15:58:34.8278435Z    Compiling matchers v0.0.1
2020-04-03T15:58:34.8278547Z    Compiling regex v1.3.3
2020-04-03T15:58:34.8278922Z    Compiling openssl-sys v0.9.53
2020-04-03T15:58:34.8279210Z    Compiling libz-sys v1.0.25
2020-04-03T15:58:34.8279471Z    Compiling backtrace-sys v0.1.32
2020-04-03T15:58:34.8279729Z    Compiling libssh2-sys v0.2.14
2020-04-03T15:58:34.8279985Z    Compiling libgit2-sys v0.10.0
2020-04-03T15:58:34.8280186Z    Compiling darling_core v0.10.2
2020-04-03T15:58:34.8280301Z    Compiling synstructure v0.12.3
2020-04-03T15:58:34.8280415Z    Compiling serde_derive v1.0.104
2020-04-03T15:58:34.8280515Z    Compiling gumdrop_derive v0.7.0
2020-04-03T15:58:34.8280802Z    Compiling thiserror-impl v1.0.10
2020-04-03T15:58:34.8281069Z    Compiling tracing-attributes v0.1.6
2020-04-03T15:58:34.8281186Z    Compiling darling_macro v0.10.2
2020-04-03T15:58:34.8281304Z    Compiling gumdrop v0.7.0
2020-04-03T15:58:34.8281418Z    Compiling thiserror v1.0.10
2020-04-03T15:58:34.8281531Z    Compiling tracing v0.1.12
2020-04-03T15:58:34.8281631Z    Compiling darling v0.10.2
2020-04-03T15:58:34.8281875Z    Compiling abscissa_derive v0.5.0
2020-04-03T15:58:34.8282001Z    Compiling toml v0.5.6
2020-04-03T15:58:34.8282118Z    Compiling chrono v0.4.10
2020-04-03T15:58:34.8282233Z    Compiling semver v0.9.0
2020-04-03T15:58:34.8282351Z    Compiling serde_json v1.0.46
2020-04-03T15:58:34.8282459Z    Compiling cvss v1.0.0
2020-04-03T15:58:34.8282578Z    Compiling platforms v0.2.1
2020-04-03T15:58:34.8282695Z    Compiling secrecy v0.6.0
2020-04-03T15:58:34.8282985Z    Compiling tracing-subscriber v0.1.6
2020-04-03T15:58:34.8283478Z    Compiling cargo-lock v4.0.1
2020-04-03T15:58:34.8283604Z    Compiling backtrace v0.3.42
2020-04-03T15:58:34.8283876Z    Compiling color-backtrace v0.3.0
2020-04-03T15:58:34.8283984Z    Compiling abscissa_core v0.5.2
2020-04-03T15:58:34.8284102Z    Compiling git2 v0.11.0
2020-04-03T15:58:34.8284369Z    Compiling crates-index v0.13.4
2020-04-03T15:58:34.8284492Z    Compiling rustsec v0.18.0
2020-04-03T15:58:34.8284891Z    Compiling cargo-audit v0.11.2 (/home/runner/work/dffml/dffml/examples/shouldi/tests/downloads/cargo-audit-download/cargo-audit-0.11.2)
2020-04-03T15:58:34.8285041Z     Finished release [optimized] target(s) in 6m 05s
2020-04-03T15:58:34.8285109Z 
```

What's happening here is that the build is completing but the `cargo` tool outputs to stderr even if nothing errors. Therefore, we need to check the return code to decide if we should throw an `Exception`

Reply:
@Yash-Varshney Nice job! Please just run `black` on this and then we can merge it!